File: tests/parental_consent_refused.spec.ts
import { test, expect, Page } from "@playwright/test";

let p: Page;

test("Parental consent - Consent refused", async ({ page }) => {
  p = page;

  await given_the_app_is_setup();
  await when_i_go_through_the_consent_journey();
  await then_i_see_the_consent_page();

  await when_i_refuse_consent();
  await and_i_click_continue();
  await then_i_see_the_reason_page();

  await when_i_choose_contains_gelatine_as_the_reason();
  await and_i_click_continue();
  await then_i_see_the_consent_confirm_page();
});

async function given_the_app_is_setup() {
  await p.goto("/reset");
}

async function when_i_go_through_the_consent_journey() {
  await p.goto("/sessions/1/consents/start");
  await p.getByRole("button", { name: "Start now" }).click();

  // Name
  await p.getByLabel("First name").fill("Joe");
  await p.getByLabel("Last name").fill("Test");
  await p.getByText("Yes").click();
  await p.getByLabel("Known as").fill("LittleJoeTests");
  await and_i_click_continue();

  // Date of birth
  await p.getByLabel("Day").fill("01");
  await p.getByLabel("Month").fill("01");
  await p.getByLabel("Year").fill("2010");
  await and_i_click_continue();

  // School
  await p.getByText("Yes, they go to this school").click();
  await and_i_click_continue();

  // Parent details
  await p.getByLabel("Your name").fill("Joe Senior");
  await p.getByRole("radio", { name: "Dad" }).click();
  await p.getByLabel("Email address").fill("joe.senior@example.com");
  await and_i_click_continue();
}

async function and_i_click_continue() {
  await p.getByRole("button", { name: "Continue" }).click();
}

async function then_i_see_the_consent_confirm_page() {
  await expect(p.locator("h1")).toContainText("Check your answers and confirm");
}

async function then_i_see_the_consent_page() {
  await expect(p.locator("h1")).toContainText(
    "Do you agree to them having the HPV vaccination?",
  );
}

async function when_i_refuse_consent() {
  await p.getByRole("radio", { name: "No" }).click();
}

async function then_i_see_the_reason_page() {
  await expect(p.locator("h1")).toContainText(
    "Please tell us why you do not agree",
  );
}

async function when_i_choose_contains_gelatine_as_the_reason() {
  await p.click("text=Vaccine contains gelatine from pigs");
}

File: spec/features/parental_consent_refused_spec.rb
# frozen_string_literal: true

describe "Parental consent" do
  around { |example| travel_to(Date.new(2025, 7, 31)) { example.run } }

  scenario "Refused" do
    stub_pds_search_to_return_no_patients

    given_an_hpv_programme_is_underway
    when_i_go_to_the_consent_form
    then_i_see_the_start_page

    when_i_fill_in_my_details
    then_i_see_the_consent_page

    when_i_refuse_consent
    then_i_can_check_my_answers

    when_i_confirm_my_answers
    and_i_refuse_to_answer_questions_on_ethnicity
    then_i_see_the_confirmation_page

    when_i_wait_for_the_background_jobs_to_complete
    and_i_receive_an_email_confirming_that_my_child_wont_be_vaccinated
    and_i_receive_a_text_confirming_that_my_child_wont_be_vaccinated

    when_the_nurse_checks_the_consent_responses
    then_they_see_that_the_child_has_consent_refused
  end

  def given_an_hpv_programme_is_underway
    @programme = Programme.hpv
    @team = create(:team, :with_one_nurse, programmes: [@programme])
    location = create(:school, name: "Pilot School", team: @team)
    @session =
      create(
        :session,
        :scheduled,
        team: @team,
        programmes: [@programme],
        location:
      )
    @child = create(:patient, session: @session)
  end

  def when_i_go_to_the_consent_form
    visit start_parent_interface_consent_forms_path(@session, @programme)
  end

  def then_i_see_the_start_page
    expect(page).to have_content(
      "Give or refuse consent for the HPV vaccination"
    )
  end

  def when_i_refuse_consent
    expect(page).to have_content("Do you agree")
    choose "No"
    click_on "Continue"

    expect(page).to have_content(
      "Please tell us why you do not agree to your child having the HPV vaccination"
    )
    choose "Medical reasons"
    click_on "Continue"

    expect(page).to have_content(
      "What medical reasons prevent your child from being vaccinated?"
    )
    fill_in "Give details", with: "They have a weakened immune system"
    click_on "Continue"
  end

  def when_i_fill_in_my_details
    click_on "Start now"

    expect(page).to have_content("What is your child’s name?")
    fill_in "First name", with: @child.given_name
    fill_in "Last name", with: @child.family_name
    choose "No" # Do they use a different name in school?
    click_on "Continue"

    expect(page).to have_content("What is your child’s date of birth?")
    fill_in "Day", with: @child.date_of_birth.day
    fill_in "Month", with: @child.date_of_birth.month
    fill_in "Year", with: @child.date_of_birth.year
    click_on "Continue"

    expect(page).to have_content("Confirm your child’s school")
    choose "Yes, they go to this school"
    click_on "Continue"

    expect(page).to have_content("About you")
    fill_in "Full name", with: "Jane #{@child.family_name}"
    choose "Mum" # Your relationship to the child
    fill_in "Email address", with: "jane@example.com"
    fill_in "Phone number", with: "07123456789"
    check "Tick this box if you’d like to get updates by text message"
    click_on "Continue"

    expect(page).to have_content("Phone contact method")
    choose "I do not have specific needs"
    click_on "Continue"
  end

  def then_i_see_the_consent_page
    expect(page).to have_content("Do you agree")
  end

  def then_i_can_check_my_answers
    expect(page).to have_content("Check and confirm")
  end

  def when_i_confirm_my_answers
    click_on "Confirm"
  end

  def then_i_see_the_confirmation_page
    expect(page).to have_content(
      "You’ve told us that you do not want #{@child.full_name(context: :parents)} to get the HPV vaccination at school"
    )
  end

  def when_i_wait_for_the_background_jobs_to_complete
    perform_enqueued_jobs(only: ProcessConsentFormJob)
  end

  def and_i_receive_an_email_confirming_that_my_child_wont_be_vaccinated
    expect_email_to "jane@example.com", :consent_confirmation_refused
  end

  def and_i_receive_a_text_confirming_that_my_child_wont_be_vaccinated
    expect_sms_to "07123 456789", :consent_confirmation_refused
  end

  def when_the_nurse_checks_the_consent_responses
    sign_in @team.users.first
    visit sessions_path
    click_on "Pilot School"
    within(".app-secondary-navigation") { click_on "Children" }
  end

  def then_they_see_that_the_child_has_consent_refused
    expect(page).to have_content("Parent refused")
    choose "Has a refusal"
    click_on "Update results"
    expect(page).to have_content(@child.full_name)
  end

  def and_i_refuse_to_answer_questions_on_ethnicity
    choose "No, skip the ethnicity questions"
    click_on "Continue"
  end
end
